gapis/resolve/dependencygraph2/dependency_graph_builder.go

```diff
@@ -0,0 +1,303 @@
+	debug(ctx, "  OnGet (%T %d)%v : %d", owner, owner.RefID(), frag, valueRef.RefID())
+	readEffect := ReadFragmentEffect{b.currentNodeID, frag}
+	writes, ok := b.fragmentWrites[owner.RefID()]
+	if !ok {
+		return
+	}
+	if _, ok := frag.(api.CompleteFragment); ok {
+		for writeFrag, writeNode := range writes {
+			writeEffect := WriteFragmentEffect{writeNode, writeFrag}
+			b.addDependency(writeEffect, readEffect)
+		}
+	} else if writeNode, ok := writes[frag]; ok {
+		writeEffect := WriteFragmentEffect{writeNode, frag}
+		b.addDependency(writeEffect, readEffect)
+	}
+}
+
+func (b *dependencyGraphBuilder) OnSet(ctx context.Context, owner api.Reference, frag api.Fragment, oldValueRef api.Reference, newValueRef api.Reference) {
+	debug(ctx, "  OnSet (%T %d)%v : %d → %d", owner, owner.RefID(), frag, oldValueRef.RefID(), newValueRef.RefID())
+	if _, ok := frag.(api.CompleteFragment); ok {
+		b.fragmentWrites[owner.RefID()] = map[api.Fragment]NodeID{frag: b.currentNodeID}
+	} else if writes, ok := b.fragmentWrites[owner.RefID()]; ok {
+		writes[frag] = b.currentNodeID
+	} else {
+		b.fragmentWrites[owner.RefID()] = map[api.Fragment]NodeID{frag: b.currentNodeID}
+	}
+}
+
+// OnWriteSlice is called when writing to a slice
+func (b *dependencyGraphBuilder) OnWriteSlice(ctx context.Context, slice memory.Slice) {
+	debug(ctx, "  OnWriteSlice: %v", slice)
+	b.addMemoryWrite(WriteMemEffect{b.currentNodeID, slice})
+}
+
+// OnReadSlice is called when reading from a slice
+func (b *dependencyGraphBuilder) OnReadSlice(ctx context.Context, slice memory.Slice) {
+	debug(ctx, "  OnReadSlice: %v", slice)
+	b.addMemoryRead(ReadMemEffect{b.currentNodeID, slice})
+}
+
+// observationSlice constructs a Slice from a CmdObservation
+func observationSlice(obs api.CmdObservation) memory.Slice {
+	return memory.NewSlice(obs.Range.Base, obs.Range.Base, obs.Range.Size, obs.Range.Size, obs.Pool, reflect.TypeOf(memory.Char(0)))
+}
+
+// OnWriteObs is called when a memory write observation becomes visible
+func (b *dependencyGraphBuilder) OnWriteObs(ctx context.Context, obs []api.CmdObservation) {
+	for i, o := range obs {
+		b.addObs(i, o, true)
+	}
+}
+
+// OnReadObs is called when a memory read observation becomes visible
+func (b *dependencyGraphBuilder) OnReadObs(ctx context.Context, obs []api.CmdObservation) {
+	for i, o := range obs {
+		b.addObs(i, o, false)
+	}
+}
+
+// OpenForwardDependency is called to begin a forward dependency.
+// See `StateWatcher.OpenForwardDependency` for an explanation of forward dependencies.
+func (b *dependencyGraphBuilder) OpenForwardDependency(ctx context.Context, dependencyID interface{}) {
+	debug(ctx, "  OpenForwardDependency: %v", dependencyID)
+	if _, ok := b.openForwardDependencies[dependencyID]; ok {
+		log.I(ctx, "OpenForwardDependency: Forward dependency opened multiple times before being closed. DependencyID: %v, close node: %v", dependencyID, b.currentNodeID)
```

Comment:
`StateWatcher.OpenForwardDependency` says there should be exactly one close per open, however this log line is getting called quite frequently in my testing.  Is this intended behaviour, and if so then perhaps it would be better at debug instead of info?